### PR TITLE
[AutoDiff upstream] Add `@derivative(of:)` attribute.

### PR DESCRIPTION
**Pull request by dan-zheng**
The `@derivative(of:)` attribute registers a function as a derivative of another function.
This patch adds the `@derivative(of:)` attribute definition, syntax, parsing, and printing.

Resolves TF-826.

Todos:
- Type-checking (TF-829).
- Serialization (TF-837).

### CODE REVIEW

**Comment by dan-zheng** [comment]
cc @rxwei @marcrasi @bgogul

File: include/swift/AST/Attr.def [hunk 1]
```diff
@@ -508,6 +508,11 @@ DECL_ATTR(differentiable, Differentiable,
   ABIStableToAdd | ABIBreakingToRemove | APIStableToAdd | APIBreakingToRemove,
   91)
 
+DECL_ATTR(differentiating, Differentiating,
```
Comment (lattner) [inline include/swift/AST/Attr.def]: Bikeshed language design issue - feel free to punt to the formal review process, but this attribute is textually very similar to @differentiable.  Did you consider something like `@derivativeOf` as the name of this attribute?

**Review by lattner — APPROVED** [review]
LGTM with a few changes, thanks!

File: include/swift/AST/Attr.def (hunk 1 above)
Comment (rxwei) [reply include/swift/AST/Attr.def]: I'm very open to this idea! Personally, I like `@derivative(of: ...)` and `@transpose(of: ...)` better than `@differentiating(...)` and `@transposing(...)`.

File: lib/AST/Attr.cpp [hunk 2]
```diff
@@ -864,6 +864,20 @@ bool DeclAttribute::printImpl(ASTPrinter &Printer, const PrintOptions &Options,
     break;
   }
 
+  case DAK_Differentiating: {
+    Printer.printAttrName("@differentiating");
+    Printer << '(';
+    auto *attr = cast<DifferentiatingAttr>(this);
+    auto *derivative = dyn_cast_or_null<AbstractFunctionDecl>(D);
+    Printer << attr->getOriginal().Name;
+    auto diffParamsString = getDifferentiationParametersClauseString(
```
[the remark this answers is not in the document]
Comment (rxwei) [reply lib/AST/Attr.cpp]: Completely agreed. This should definitely be `cast<>`. `getDifferentiationParametersClauseString`'s semantics are unclear regarding whether the first argument can be null -- it appears that it is dereferencing the pointer without any check at one point.

File: include/swift/AST/Attr.def (hunk 1 above)
Comment (lattner) [reply include/swift/AST/Attr.def]: +1 for `@derivative(of: ...)`

Comment (dan-zheng) [reply include/swift/AST/Attr.def]: `@derivative(of: ...)` does clearly evoke derivative registration! I'll rename to `@differentiating(...)` to `@derivative(of: ...)`, and deprecate `@differentiating(...)` on `tensorflow` branch with a warning, if there are no objections.

Comment (saeta) [reply include/swift/AST/Attr.def]: +1

File: lib/AST/Attr.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dan-zheng) [reply lib/AST/Attr.cpp]: Thanks for the suggestions! I addressed feedback in https://github.com/apple/swift/pull/28377:
* Clarified doc comments and parameter non-nullability for `getDifferentiationParametersClauseString`.
* Changed existing occurrence of `dyn_cast_or_null<AbstractFunctionDecl>(D)` to `cast<AbstractFunctionDecl>(D)`.

File: lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -1059,6 +1059,80 @@ bool Parser::parseDifferentiableAttributeArguments(
...
+  auto consumeIfTrailingComma = [&]() -> bool {
+    if (!consumeIf(tok::comma)) return false;
+    // Diagnose trailing comma before ')'.
+    if (Tok.is(tok::r_paren)) {
+      diagnose(Tok, diag::unexpected_separator, ",");
+      return true;
+    }
+    // Check that token after comma is 'linear' or 'wrt:'.
+    if (!Tok.is(tok::identifier) ||
+        !(Tok.getText() == "linear" || Tok.getText() == "wrt")) {
```
[the remark this answers is not in the document]
Comment (dan-zheng) [reply lib/Parse/ParseDecl.cpp]: > This logic is very difficult to read with all the negation happening, I'd recommend inverting this to be "if (positive test) return false; emit error"

Refactored existing code to use simpler positive conditions in https://github.com/apple/swift/pull/28377. I'll update this PR too.

File: include/swift/AST/Attr.def (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/Attr.def]: Renamed `@differentiating` to `@derivative(of:)` in https://github.com/apple/swift/pull/28321/commits/4b9f025351afef770a431ae276581d5cf473e736.

File: lib/Parse/ParseDecl.cpp [hunk 4]
```diff
@@ -1059,6 +1059,80 @@ bool Parser::parseDifferentiableAttributeArguments(
   return false;
 }
 
+ParserResult<DifferentiatingAttr>
```
[the remark this answers is not in the document]
Comment (dan-zheng) [reply lib/Parse/ParseDecl.cpp]: Done.

Also added production rules for `Parser::parseDifferentiationParametersClause` and `Parser::parseDifferentiableAttribute`.

**Comment by dan-zheng** [comment]
I addressed all comments and rebased onto master.

There are no major changes: the biggest syntax change is renaming `@differentiating` to `@derivative(of:)`, which has already been done on `tensorflow` branch: https://github.com/apple/swift/pull/28481.

**Comment by dan-zheng** [comment]
@swift-ci Please test
https://github.com/apple/swift-syntax/pull/188

**Comment by dan-zheng** [comment]
@swift-ci Please test
apple/swift-syntax#188

**Comment by dan-zheng** [comment]
apple/swift-syntax#188
@swift-ci Please test

**Comment by dan-zheng** [comment]
Merging since there are no major changes since the last approval and CI has passed.

**Pull request merged**